File: AcademicInfo/src/main/kotlin/domain/Faculty.kt
package domain

import org.ktorm.entity.Entity
import org.ktorm.schema.Table
import org.ktorm.schema.int
import org.ktorm.schema.text

interface Faculty : Entity<Faculty> {
    val facultyId : Int
    val facultyName : String
    val address : String
    val email : String
    val phoneNumber : String
    val chiefTeacherId : Teacher
}

object Faculties : Table<Faculty>("Faculty") {
    val facultyId = int("FacultyID").primaryKey().bindTo { it.facultyId }
    val facultyName = text("FacultyName").bindTo { it.facultyName }
    val address = text("Address").bindTo { it.address }
    val email = text("Email").bindTo { it.email }
    val phoneNumber = text("PhoneNumber").bindTo { it.phoneNumber }
    val chiefTeacherId = int("ChiefTeacherID").references(Teachers) { it.chiefTeacherId }
}

File: academic-info-golang/model/faculty.go
package model

type Faculty struct {
	FacultyId      int    `gorm:"primaryKey;column:faculty_id;type:int;"`
	FacultyName    string `gorm:"column:faculty_name;type:string"`
	Email          string `gorm:"column:email;type:string"`
	PhoneNumber    string `gorm:"column:phone_number;type:string"`
	Address        string `gorm:"column:address;type:string"`
	ChiefTeacherId int

	// This is actually the chief teacher of the faculty,
	// but GORM has a naming convention that prevents me from naming it ChiefTeacher.
	Teacher *Teacher `gorm:"foreignKey:ChiefTeacherId;references:TeacherId;"`
}

func (Faculty) TableName() string {
	return "faculty"
}
